rebuild/src/signature.h

```diff
@@ -0,0 +1,119 @@
+    virtual void AddTo(blst::P2 &point) = 0;
+    virtual blst::P2 MultiplyBy(
+        blst::byte *rand_bytes, size_t rand_bytes_length) = 0;
+    virtual P2AffineGroup AsAffine() = 0;
+    virtual void Sign(
+        const blst::SecretKey &key,
+        uint8_t *msg,
+        size_t msg_length,
+        const std::string &dst) = 0;
+};
+
+class P2 : public P2Wrapper {
+    blst::P2 _point;
+
+   public:
+    P2(blst::P2 point) : _point(std::move(point)) {}
+    bool IsInfinite() override { return _point.is_inf(); }
+    bool InGroup() override { return _point.in_group(); }
+    void Serialize(bool compress, blst::byte *out) override {
+        compress ? _point.compress(out) : _point.serialize(out);
+    }
+    void AddTo(blst::P2 &point) override { point.add(_point); };
+    blst::P2 MultiplyBy(
+        blst::byte *rand_bytes, size_t rand_bytes_length) override {
+        blst::byte out[192];
+        _point.serialize(out);
+        // this should get std::move all the way into the P2 member value
+        blst::P2 point{out, 192};
+        point.mult(rand_bytes, rand_bytes_length);
+        return point;
+    };
+    P2AffineGroup AsAffine() override {
+        P2AffineGroup group{std::make_unique<blst::P2_Affine>(_point), nullptr};
+        group.raw_point = group.smart_pointer.get();
+        return group;
+    };
+    void Sign(
+        const blst::SecretKey &key,
+        uint8_t *msg,
+        size_t msg_length,
+        const std::string &dst) override {
+        _point.hash_to(msg, msg_length, dst);
+        _point.sign_with(key);
+    };
+};
+
+class P2Affine : public P2Wrapper {
+    blst::P2_Affine _point;
+
+   public:
+    P2Affine(blst::P2_Affine point) : _point(std::move(point)) {}
+    bool IsInfinite() override { return _point.is_inf(); }
+    bool InGroup() override { return _point.in_group(); }
+    void Serialize(bool compress, blst::byte *out) override {
+        compress ? _point.compress(out) : _point.serialize(out);
+    }
+    void AddTo(blst::P2 &point) override { point.add(_point); };
+    blst::P2 MultiplyBy(
+        blst::byte *rand_bytes, size_t rand_bytes_length) override {
+        blst::byte out[192];
+        _point.serialize(out);
+        // this should get std::move all the way into the P2 member value
+        blst::P2 point{out, 192};
+        point.mult(rand_bytes, rand_bytes_length);
+        return point;
+    };
+    P2AffineGroup AsAffine() override {
+        P2AffineGroup group{nullptr, &_point};
+        return group;
+    }
+    void Sign(
+        const blst::SecretKey &key,
+        uint8_t *msg,
+        size_t msg_length,
+        const std::string &dst) override {
+        blst::P2 jacobian{_point};
+        jacobian.hash_to(msg, msg_length, dst);
+        jacobian.sign_with(key);
+        _point = jacobian.to_affine();
+    };
+};
+
+class Signature : public Napi::ObjectWrap<Signature> {
+   public:
+    std::unique_ptr<P2Wrapper> _point;
```

Comment:
should `_point` be private? Also, might make sense to explicitly initialize it to `nullptr`.

Reply:
They are polymorphic wrappers around P1 and P1Affine (and P2 and P2Affine in Signature) to allow polymorphic access to the underlying methods of the two different points.  They are used in lots of places in the code and cannot, and I would say probably should not, be made private.

Reply:
I'm concerned that a pointer made public can be invalidated virtually everywhere, making it difficult to pinpoint potential error culprits. Regardless, if you don't intend to do any encapsulation here (which is completely fine), I would suggest changing `class` to `struct` to hint that it's all publicly accessible (and it is, even with `class` as everything is `public`). Then, `_point` should be stripped of `_` prefix to hint that its usage is not messing with internals and intended. Does it make sense?

Reply:
OK.  I went with removing the underscore prefix. They need to be classes because they extend `Napi::ObjectWrap`.  Fixed in bf0461f5a8d0a6db403ae4fdcf9af4cf925a69b7